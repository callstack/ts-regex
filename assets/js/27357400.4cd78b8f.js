"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[119],{4061:e=>{e.exports=JSON.parse('{"name":"docusaurus-plugin-content-pages","id":"default"}')}}]);