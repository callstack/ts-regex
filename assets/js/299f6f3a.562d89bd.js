"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[425],{6870:(e,n,r)=>{r.r(n),r.d(n,{assets:()=>o,contentTitle:()=>c,default:()=>u,frontMatter:()=>a,metadata:()=>i,toc:()=>d});var t=r(4848),s=r(8453);const a={id:"captures",title:"Captures"},c=void 0,i={id:"api/captures",title:"Captures",description:"capture()",source:"@site/docs/api/captures.md",sourceDirName:"api",slug:"/api/captures",permalink:"/ts-regex-builder/api/captures",draft:!1,unlisted:!1,tags:[],version:"current",frontMatter:{id:"captures",title:"Captures"},sidebar:"docs",previous:{title:"Constructs",permalink:"/ts-regex-builder/api/constructs"},next:{title:"Quantifiers",permalink:"/ts-regex-builder/api/quantifiers"}},o={},d=[{value:"<code>capture()</code>",id:"capture",level:3},{value:"Named groups",id:"named-groups",level:4},{value:"<code>ref()</code>",id:"ref",level:3}];function l(e){const n={a:"a",admonition:"admonition",code:"code",h3:"h3",h4:"h4",li:"li",p:"p",pre:"pre",ul:"ul",...(0,s.R)(),...e.components};return(0,t.jsxs)(t.Fragment,{children:[(0,t.jsx)(n.h3,{id:"capture",children:(0,t.jsx)(n.code,{children:"capture()"})}),"\n",(0,t.jsx)(n.pre,{children:(0,t.jsx)(n.code,{className:"language-ts",children:"function capture(\n  sequence: RegexSequence,\n  options?: {\n    name?: string;\n  },\n): Capture;\n"})}),"\n",(0,t.jsx)(n.p,{children:"Regex syntax:"}),"\n",(0,t.jsxs)(n.ul,{children:["\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"(...)"})," for capturing groups (no ",(0,t.jsx)(n.code,{children:"name"})," option)"]}),"\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"(?<name>...)"})," for named capturing groups (",(0,t.jsx)(n.code,{children:"name"})," option)"]}),"\n"]}),"\n",(0,t.jsx)(n.p,{children:"Captures, also known as capturing groups, extract and store parts of the matched string for later use."}),"\n",(0,t.jsxs)(n.p,{children:["Capture results are available using array-like ",(0,t.jsxs)(n.a,{href:"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match#using_match",children:[(0,t.jsx)(n.code,{children:"match()"})," result object"]}),"."]}),"\n",(0,t.jsx)(n.h4,{id:"named-groups",children:"Named groups"}),"\n",(0,t.jsxs)(n.p,{children:["When using ",(0,t.jsx)(n.code,{children:"name"})," options, the group becomes a ",(0,t.jsx)(n.a,{href:"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Named_capturing_group",children:"named capturing group"})," allowing to refer to it using name instead of index."]}),"\n",(0,t.jsxs)(n.p,{children:["Named capture results are available using ",(0,t.jsx)(n.a,{href:"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match#using_named_capturing_groups",children:(0,t.jsx)(n.code,{children:"groups"})})," property on ",(0,t.jsx)(n.code,{children:"match()"})," result."]}),"\n",(0,t.jsx)(n.admonition,{type:"note",children:(0,t.jsxs)(n.p,{children:["TS Regex Builder does not have a construct for non-capturing groups. Such groups are implicitly added when required. E.g., ",(0,t.jsx)(n.code,{children:'zeroOrMore("abc")'})," is encoded as ",(0,t.jsx)(n.code,{children:"(?:abc)+"}),"."]})}),"\n",(0,t.jsx)(n.h3,{id:"ref",children:(0,t.jsx)(n.code,{children:"ref()"})}),"\n",(0,t.jsx)(n.pre,{children:(0,t.jsx)(n.code,{className:"language-ts",children:"function ref(name: string): Reference;\n"})}),"\n",(0,t.jsxs)(n.p,{children:["Regex syntax: ",(0,t.jsx)(n.code,{children:"\\k<...>"}),"."]}),"\n",(0,t.jsx)(n.p,{children:"Creates a reference, also known as a backreference, which allows matching again the exact text that a capturing group previously matched. The reference must use the same name as some capturing group earlier in the expression to form a valid regex."}),"\n",(0,t.jsxs)(n.p,{children:["Usage with ",(0,t.jsx)(n.code,{children:"capture()"}),":"]}),"\n",(0,t.jsx)(n.pre,{children:(0,t.jsx)(n.code,{className:"language-ts",children:"const regex = buildRegExp([\n  // Create a named capture using name from `someKey`.\n  capture(..., { name: 'someKey' }),\n  // ... some other elements ...\n\n  // Match the same text as matched by `capture` with the same name.\n  ref('someKey'),\n  ])\n"})}),"\n",(0,t.jsx)(n.admonition,{type:"note",children:(0,t.jsxs)(n.p,{children:["TS Regex Builder doesn't support using ordinal backreferences (",(0,t.jsx)(n.code,{children:"\\1"}),", ",(0,t.jsx)(n.code,{children:"\\2"}),", etc) because in complex regex patterns, these references are difficult to accurately use."]})})]})}function u(e={}){const{wrapper:n}={...(0,s.R)(),...e.components};return n?(0,t.jsx)(n,{...e,children:(0,t.jsx)(l,{...e})}):l(e)}},8453:(e,n,r)=>{r.d(n,{R:()=>c,x:()=>i});var t=r(6540);const s={},a=t.createContext(s);function c(e){const n=t.useContext(a);return t.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function i(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:c(e.components),t.createElement(a.Provider,{value:n},e.children)}}}]);