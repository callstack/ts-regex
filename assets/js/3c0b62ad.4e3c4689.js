"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[544],{6639:(e,n,r)=>{r.r(n),r.d(n,{assets:()=>a,contentTitle:()=>c,default:()=>l,frontMatter:()=>o,metadata:()=>i,toc:()=>d});var t=r(4848),s=r(8453);const o={id:"anchors",title:"Anchors",sidebar_position:6},c=void 0,i={id:"api/anchors",title:"Anchors",description:"Anchors are special characters or sequences that specify positions in the input string rather than matching specific characters.",source:"@site/docs/api/anchors.md",sourceDirName:"api",slug:"/api/anchors",permalink:"/docs/api/anchors",draft:!1,unlisted:!1,tags:[],version:"current",sidebarPosition:6,frontMatter:{id:"anchors",title:"Anchors",sidebar_position:6},sidebar:"docs",previous:{title:"Character Classes",permalink:"/docs/api/character-classes"},next:{title:"Examples",permalink:"/docs/examples"}},a={},d=[{value:"Start and end of string",id:"start-and-end-of-string",level:3},{value:"Word boundary",id:"word-boundary",level:3}];function h(e){const n={code:"code",h3:"h3",li:"li",p:"p",pre:"pre",ul:"ul",...(0,s.R)(),...e.components};return(0,t.jsxs)(t.Fragment,{children:[(0,t.jsx)(n.p,{children:"Anchors are special characters or sequences that specify positions in the input string rather than matching specific characters."}),"\n",(0,t.jsx)(n.h3,{id:"start-and-end-of-string",children:"Start and end of string"}),"\n",(0,t.jsx)(n.pre,{children:(0,t.jsx)(n.code,{className:"language-ts",children:"const startOfString: Anchor;\nconst endOfString: Anchor;\n"})}),"\n",(0,t.jsxs)(n.ul,{children:["\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"startOfString"})," anchor matches the start of a string (or line, if multiline mode is enabled). Regex syntax: ",(0,t.jsx)(n.code,{children:"^"}),"."]}),"\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"endOfString"})," anchor matches the end of a string (or line, if multiline mode is enabled). Regex syntax: ",(0,t.jsx)(n.code,{children:"$"}),"."]}),"\n"]}),"\n",(0,t.jsx)(n.h3,{id:"word-boundary",children:"Word boundary"}),"\n",(0,t.jsx)(n.pre,{children:(0,t.jsx)(n.code,{className:"language-ts",children:"const wordBoundary: Anchor;\nconst notWordBoundary: Anchor;\n"})}),"\n",(0,t.jsxs)(n.ul,{children:["\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"wordBoundary"})," matches the positions where a word character is not followed or preceded by another word character, effectively indicating the start or end of a word. Regex syntax: ",(0,t.jsx)(n.code,{children:"\\b"}),"."]}),"\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"notWordBoundary"})," matches the positions where a word character is followed or preceded by another word character, indicating that it is not at the start or end of a word. Regex syntax: ",(0,t.jsx)(n.code,{children:"\\B"}),"."]}),"\n"]}),"\n",(0,t.jsxs)(n.p,{children:["Note: word characters are letters, digits, and underscore (",(0,t.jsx)(n.code,{children:"_"}),"). Other special characters like ",(0,t.jsx)(n.code,{children:"#"}),", ",(0,t.jsx)(n.code,{children:"$"}),", etc are not considered word characters."]})]})}function l(e={}){const{wrapper:n}={...(0,s.R)(),...e.components};return n?(0,t.jsx)(n,{...e,children:(0,t.jsx)(h,{...e})}):h(e)}},8453:(e,n,r)=>{r.d(n,{R:()=>c,x:()=>i});var t=r(6540);const s={},o=t.createContext(s);function c(e){const n=t.useContext(o);return t.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function i(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:c(e.components),t.createElement(o.Provider,{value:n},e.children)}}}]);