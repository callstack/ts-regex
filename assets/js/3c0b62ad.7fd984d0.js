"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[544],{6639:(e,n,r)=>{r.r(n),r.d(n,{assets:()=>a,contentTitle:()=>o,default:()=>l,frontMatter:()=>i,metadata:()=>c,toc:()=>d});var t=r(4848),s=r(8453);const i={id:"anchors",title:"Anchors",sidebar_position:6},o=void 0,c={id:"api/anchors",title:"Anchors",description:"Anchors are special characters or sequences that specify positions in the input string rather than matching specific characters.",source:"@site/docs/api/anchors.md",sourceDirName:"api",slug:"/api/anchors",permalink:"/ts-regex-builder/api/anchors",draft:!1,unlisted:!1,tags:[],version:"current",sidebarPosition:6,frontMatter:{id:"anchors",title:"Anchors",sidebar_position:6},sidebar:"docs",previous:{title:"Character Classes",permalink:"/ts-regex-builder/api/character-classes"},next:{title:"Examples",permalink:"/ts-regex-builder/examples"}},a={},d=[{value:"Start and end of string",id:"start-and-end-of-string",level:3},{value:"Word boundary",id:"word-boundary",level:3}];function h(e){const n={code:"code",h3:"h3",li:"li",p:"p",pre:"pre",ul:"ul",...(0,s.R)(),...e.components};return(0,t.jsxs)(t.Fragment,{children:[(0,t.jsx)(n.p,{children:"Anchors are special characters or sequences that specify positions in the input string rather than matching specific characters."}),"\n",(0,t.jsx)(n.h3,{id:"start-and-end-of-string",children:"Start and end of string"}),"\n",(0,t.jsx)(n.pre,{children:(0,t.jsx)(n.code,{className:"language-ts",children:"const startOfString: Anchor;\nconst endOfString: Anchor;\n"})}),"\n",(0,t.jsxs)(n.ul,{children:["\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"startOfString"})," anchor matches the start of a string (or line, if multiline mode is enabled). Regex syntax: ",(0,t.jsx)(n.code,{children:"^"}),"."]}),"\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"endOfString"})," anchor matches the end of a string (or line, if multiline mode is enabled). Regex syntax: ",(0,t.jsx)(n.code,{children:"$"}),"."]}),"\n"]}),"\n",(0,t.jsx)(n.h3,{id:"word-boundary",children:"Word boundary"}),"\n",(0,t.jsx)(n.pre,{children:(0,t.jsx)(n.code,{className:"language-ts",children:"const wordBoundary: Anchor;\nconst notWordBoundary: Anchor;\n"})}),"\n",(0,t.jsxs)(n.ul,{children:["\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"wordBoundary"})," matches the positions where a word character is not followed or preceded by another word character, effectively indicating the start or end of a word. Regex syntax: ",(0,t.jsx)(n.code,{children:"\\b"}),"."]}),"\n",(0,t.jsxs)(n.li,{children:[(0,t.jsx)(n.code,{children:"notWordBoundary"})," matches the positions where a word character is followed or preceded by another word character, indicating that it is not at the start or end of a word. Regex syntax: ",(0,t.jsx)(n.code,{children:"\\B"}),"."]}),"\n"]}),"\n",(0,t.jsxs)(n.p,{children:["Note: word characters are letters, digits, and underscore (",(0,t.jsx)(n.code,{children:"_"}),"). Other special characters like ",(0,t.jsx)(n.code,{children:"#"}),", ",(0,t.jsx)(n.code,{children:"$"}),", etc are not considered word characters."]})]})}function l(e={}){const{wrapper:n}={...(0,s.R)(),...e.components};return n?(0,t.jsx)(n,{...e,children:(0,t.jsx)(h,{...e})}):h(e)}},8453:(e,n,r)=>{r.d(n,{R:()=>o,x:()=>c});var t=r(6540);const s={},i=t.createContext(s);function o(e){const n=t.useContext(i);return t.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function c(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:o(e.components),t.createElement(i.Provider,{value:n},e.children)}}}]);