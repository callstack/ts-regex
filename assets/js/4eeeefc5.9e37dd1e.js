"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[421],{5631:(e,c,s)=>{s.r(c),s.d(c,{assets:()=>i,contentTitle:()=>t,default:()=>o,frontMatter:()=>r,metadata:()=>l,toc:()=>d});var a=s(4848),n=s(8453);const r={id:"character-classes",title:"Character Classes"},t=void 0,l={id:"api/character-classes",title:"Character Classes",description:"Character classes are a set of characters that match any one of the characters in the set.",source:"@site/docs/api/character-classes.md",sourceDirName:"api",slug:"/api/character-classes",permalink:"/ts-regex-builder/api/character-classes",draft:!1,unlisted:!1,tags:[],version:"current",frontMatter:{id:"character-classes",title:"Character Classes"},sidebar:"docs",previous:{title:"Quantifiers",permalink:"/ts-regex-builder/api/quantifiers"},next:{title:"Assertions",permalink:"/ts-regex-builder/api/assertions"}},i={},d=[{value:"Common character classes",id:"common-character-classes",level:3},{value:"<code>anyOf()</code>",id:"anyof",level:3},{value:"<code>charRange()</code>",id:"charrange",level:3},{value:"<code>charClass()</code>",id:"charclass",level:3},{value:"<code>negated()</code>",id:"negated",level:3}];function h(e){const c={code:"code",h3:"h3",li:"li",p:"p",pre:"pre",strong:"strong",ul:"ul",...(0,n.R)(),...e.components};return(0,a.jsxs)(a.Fragment,{children:[(0,a.jsx)(c.p,{children:"Character classes are a set of characters that match any one of the characters in the set."}),"\n",(0,a.jsx)(c.h3,{id:"common-character-classes",children:"Common character classes"}),"\n",(0,a.jsx)(c.pre,{children:(0,a.jsx)(c.code,{className:"language-ts",children:"const any: CharacterClass;\nconst word: CharacterClass;\nconst nonWord: CharacterClass;\nconst digit: CharacterClass;\nconst nonDigit: CharacterClass;\nconst whitespace: CharacterClass;\nconst nonWhitespace: CharacterClass;\n"})}),"\n",(0,a.jsxs)(c.ul,{children:["\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"any"})," matches any character except newline characters. Regex syntax: ",(0,a.jsx)(c.code,{children:"*"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"word"})," matches any word character (letters, digits & underscore). Regex syntax: ",(0,a.jsx)(c.code,{children:"\\w"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"nonWord"})," matches any character ",(0,a.jsx)(c.strong,{children:"except"})," word characters (letters, digits & underscore). Regex syntax: ",(0,a.jsx)(c.code,{children:"\\W"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"digit"})," matches any digit. Regex syntax: ",(0,a.jsx)(c.code,{children:"\\d"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"nonDigit"})," matches any character ",(0,a.jsx)(c.strong,{children:"except"})," digits. Regex syntax: ",(0,a.jsx)(c.code,{children:"\\D"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"whitespace"})," matches any whitespace character (spaces, tabs, line breaks). Regex syntax: ",(0,a.jsx)(c.code,{children:"\\s"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"nonWhitespace"})," matches any character ",(0,a.jsx)(c.strong,{children:"except"})," whitespace characters (spaces, tabs, line breaks). Regex syntax: ",(0,a.jsx)(c.code,{children:"\\S"}),"."]}),"\n"]}),"\n",(0,a.jsx)(c.h3,{id:"anyof",children:(0,a.jsx)(c.code,{children:"anyOf()"})}),"\n",(0,a.jsx)(c.pre,{children:(0,a.jsx)(c.code,{className:"language-ts",children:"function anyOf(characters: string): CharacterClass;\n"})}),"\n",(0,a.jsxs)(c.p,{children:["Regex syntax: ",(0,a.jsx)(c.code,{children:"[abc]"}),"."]}),"\n",(0,a.jsxs)(c.p,{children:["The ",(0,a.jsx)(c.code,{children:"anyOf"})," class matches any character in the ",(0,a.jsx)(c.code,{children:"character"})," string."]}),"\n",(0,a.jsxs)(c.p,{children:["Example: ",(0,a.jsx)(c.code,{children:"anyOf('aeiou')"})," will match either ",(0,a.jsx)(c.code,{children:"a"}),", ",(0,a.jsx)(c.code,{children:"e"}),", ",(0,a.jsx)(c.code,{children:"i"})," ",(0,a.jsx)(c.code,{children:"o"})," or ",(0,a.jsx)(c.code,{children:"u"})," characters."]}),"\n",(0,a.jsx)(c.h3,{id:"charrange",children:(0,a.jsx)(c.code,{children:"charRange()"})}),"\n",(0,a.jsx)(c.pre,{children:(0,a.jsx)(c.code,{className:"language-ts",children:"function charRange(start: string, end: string): CharacterClass;\n"})}),"\n",(0,a.jsxs)(c.p,{children:["Regex syntax: ",(0,a.jsx)(c.code,{children:"[a-z]"}),"."]}),"\n",(0,a.jsxs)(c.p,{children:["The ",(0,a.jsx)(c.code,{children:"charRange"})," class matches any characters in the range from ",(0,a.jsx)(c.code,{children:"start"})," to ",(0,a.jsx)(c.code,{children:"end"})," (inclusive)."]}),"\n",(0,a.jsx)(c.p,{children:"Examples:"}),"\n",(0,a.jsxs)(c.ul,{children:["\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"charRange('a', 'z')"})," will match all lowercase characters from ",(0,a.jsx)(c.code,{children:"a"})," to ",(0,a.jsx)(c.code,{children:"z"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"charRange('A', 'Z')"})," will match all uppercase characters from ",(0,a.jsx)(c.code,{children:"A"})," to ",(0,a.jsx)(c.code,{children:"Z"}),"."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"charRange('0', '9')"})," will match all digit characters from ",(0,a.jsx)(c.code,{children:"0"})," to ",(0,a.jsx)(c.code,{children:"9"}),"."]}),"\n"]}),"\n",(0,a.jsx)(c.h3,{id:"charclass",children:(0,a.jsx)(c.code,{children:"charClass()"})}),"\n",(0,a.jsx)(c.pre,{children:(0,a.jsx)(c.code,{className:"language-ts",children:"function charClass(...elements: CharacterClass[]): CharacterClass;\n"})}),"\n",(0,a.jsxs)(c.p,{children:["Regex syntax: ",(0,a.jsx)(c.code,{children:"[...]"}),"."]}),"\n",(0,a.jsxs)(c.p,{children:["The ",(0,a.jsx)(c.code,{children:"charClass"})," construct creates a new character class that includes all passed character classes."]}),"\n",(0,a.jsx)(c.p,{children:"Examples:"}),"\n",(0,a.jsxs)(c.ul,{children:["\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"charClass(charRange('a', 'f'), digit)"})," will match all lowercase hex digits (",(0,a.jsx)(c.code,{children:"0"})," to ",(0,a.jsx)(c.code,{children:"9"})," and ",(0,a.jsx)(c.code,{children:"a"})," to ",(0,a.jsx)(c.code,{children:"f"}),")."]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"charClass(charRange('a', 'z'), digit, anyOf(\"._-\"))"})," will match any digit, lowercase Latin letter from ",(0,a.jsx)(c.code,{children:"a"})," to ",(0,a.jsx)(c.code,{children:"z"}),", and either of ",(0,a.jsx)(c.code,{children:"."}),", ",(0,a.jsx)(c.code,{children:"_"}),", and ",(0,a.jsx)(c.code,{children:"-"})," characters."]}),"\n"]}),"\n",(0,a.jsx)(c.h3,{id:"negated",children:(0,a.jsx)(c.code,{children:"negated()"})}),"\n",(0,a.jsx)(c.pre,{children:(0,a.jsx)(c.code,{className:"language-ts",children:"function negated(element: CharacterClass): CharacterClass;\n"})}),"\n",(0,a.jsxs)(c.p,{children:["Regex syntax: ",(0,a.jsx)(c.code,{children:"[^...]"}),"."]}),"\n",(0,a.jsxs)(c.p,{children:["The ",(0,a.jsx)(c.code,{children:"negated"})," construct creates a new character class that matches any character not present in the passed character class."]}),"\n",(0,a.jsx)(c.p,{children:"Examples:"}),"\n",(0,a.jsxs)(c.ul,{children:["\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"negated(digit)"})," matches any character that is not a digit"]}),"\n",(0,a.jsxs)(c.li,{children:[(0,a.jsx)(c.code,{children:"negated(anyOf('aeiou'))"})," matches any character that is not a lowercase vowel."]}),"\n"]})]})}function o(e={}){const{wrapper:c}={...(0,n.R)(),...e.components};return c?(0,a.jsx)(c,{...e,children:(0,a.jsx)(h,{...e})}):h(e)}},8453:(e,c,s)=>{s.d(c,{R:()=>t,x:()=>l});var a=s(6540);const n={},r=a.createContext(n);function t(e){const c=a.useContext(r);return a.useMemo((function(){return"function"==typeof e?e(c):{...c,...e}}),[c,e])}function l(e){let c;return c=e.disableParentContext?"function"==typeof e.components?e.components(n):e.components||n:t(e.components),a.createElement(r.Provider,{value:c},e.children)}}}]);