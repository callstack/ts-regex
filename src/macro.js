import { createMacro, MacroError } from 'babel-plugin-macros';
import pkg from '../package.json';
import { buildPattern, capture } from '.';

// `createMacro` is simply a function that ensures your macro is only
// called in the context of a babel transpilation and will throw an
// error with a helpful message if someone does not have babel-plugin-macros
// configured correctly
module.exports = createMacro(tsRegexMacro);

function tsRegexMacro({ references, state, babel }) {
  console.log('References: ', references);

  references.buildRegExp.forEach(({ parentPath: buildRegExpCall }) => {
    if (!buildRegExpCall.isCallExpression()) {
      throw new MacroError(
        `${pkg.name}/macro should be used as function call, instead you have used it as part of ${regexgenCall.node.type}.`,
      );
    }

    //console.log('BuildRegexCall', buildRegExpCall);

    const args = buildRegExpCall.get('arguments');
    const regexAst = handleRegexSequence(args[0]);
    const pattern = buildPattern(regexAst);
    console.log('VALUE', pattern);
    buildRegExpCall.replaceWith(babel.types.regExpLiteral(pattern));
  });
}

function handleRegexSequence(path) {
  if (path.isArrayExpression()) {
    //console.log('Sequence.Array', path.get('elements'));
    return path.get('elements').map(handleRegexElement).join('');
  }

  //console.log('Sequence.Single', path);
  return handleRegexElement(path);
}

function handleRegexElement(path) {
  if (path.isStringLiteral()) {
    //console.log('Element.StringLiteral', path);
    return path.node.value;
  }

  if (path.isCallExpression()) {
    console.log('Element.CallExpression');

    const callee = path.get('callee');
    const calleeName = callee.node.name;
    const args = callee.get('arguments');
    switch (calleeName) {
      case 'capture':
        return capture(handleRegexSequence(args));
      default:
        throw new MacroError(
          `${pkg.name} "handleRegexElement" can't handle CallExpression "${calleeName}".`,
        );
    }
  }

  console.log('Unknown element:', path);
  throw new MacroError(`${pkg.name} "handleRegexElement" can't handle ${path.node?.type}.`);
}
