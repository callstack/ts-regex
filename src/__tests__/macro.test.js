import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: [
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp('a');
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp('abc');
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp(['d', 'e', 'f']);
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp(['d', 'e', 'f']);
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp(capture('a'));
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp(zeroOrMore('a'));
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp(oneOrMore('a'));
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp(optional('a'));
    `,
    `
    import { buildRegExp } from '../macro.js'
    buildRegExp(choiceOf('a','b'));
    `,
  ],
});
