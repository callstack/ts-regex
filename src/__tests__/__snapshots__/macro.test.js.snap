// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros 1. macros: 1. macros 1`] = `


import { buildRegExp } from '../macro.js'
buildRegExp('a');


      ↓ ↓ ↓ ↓ ↓ ↓

/a/;

`;

exports[`macros 2. macros: 2. macros 1`] = `


import { buildRegExp } from '../macro.js'
buildRegExp('abc');


      ↓ ↓ ↓ ↓ ↓ ↓

/abc/;

`;

exports[`macros 3. macros: 3. macros 1`] = `


import { buildRegExp } from '../macro.js'
buildRegExp(['d', 'e', 'f']);


      ↓ ↓ ↓ ↓ ↓ ↓

/def/;

`;

exports[`macros 4. macros: 4. macros 1`] = `


import { buildRegExp } from '../macro.js'
buildRegExp(['d', 'e', 'f']);


      ↓ ↓ ↓ ↓ ↓ ↓

/def/;

`;
